File: sorting/mergesort/mergesort.c

#include <assert.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

int *_mergesort(int *input, int count)
{
    /* will not happen during recursion. */
    if (input == NULL || count == 0) {
        return NULL;
    }

    if (count < 2) {
        /* XXX: return the array we received */
        int *t = malloc(sizeof(int));
        t[0] = input[0];
        return t;
    }

    /* split into two pieces. */
    int left = count / 2;
    int right = count - left;

    int *a = _mergesort(input, left);
    int *b = _mergesort(input+left, right);

    int *c = malloc(sizeof(int) * count);

    int i = 0;
    int ai = 0;
    int bi = 0;

    while (i < count)
    {
        if (ai < left && bi < right)
        {
            if (a[ai] < b[bi])
            {
                c[i] = a[ai];
                ai++;
            }
            else
            {
                c[i] = b[bi];
                bi++;
            }
        }
        else if (ai < left)
        {
            c[i] = a[ai];
            ai++;
        }
        else
        {
            c[i] = b[bi];
            bi++;
        }

        i++;
    }

    free(a);
    free(b);

    /* higher call frees this, just like above. */
    return c;
}

File: sorting/mergesort/mergesort.py
# mergesort prototype

from json import dumps

def mergesort(input, count):

    if count < 2:
        return input

    al = count / 2
    bl = count - al

    a = mergesort(input[:al], al)
    b = mergesort(input[al:], bl)

    c = []

    ai = 0
    bi = 0
    i = 0

    # do the actual sorting.
    while i < count:
        if ai < len(a) and bi < len(b):
            if a[ai] > b[bi]:
                c.append(b[bi])
                bi += 1
            else:
                c.append(a[ai])
                ai += 1
        elif ai < len(a):
            c.append(a[ai])
            ai += 1
        elif bi < len(b):
            c.append(b[bi])
            bi += 1
        i += 1

    return c

def main():

    initial_list = (0, 3, 8, 6, 4, 10, 9, 15, 11)

    print dumps(initial_list)
    print dumps(mergesort(initial_list, len(initial_list)))

    return

if __name__ == "__main__":
    main()
